### ISSUE TITLE
Set network info with ipconfig0

### ISSUE THREAD

**Issue by AcidCrash1337** [Telmate/terraform-provider-proxmox#272]
I've been working on this for about 6 hours now, repeatedly rebuilding my base image (ubuntu20.04) with different options to try to figure out what I need to do to get ipconfig0 to set my IP address.

The key line in my preseed file is this:
```
d-i pkgsel/include string open-vm-tools openssh-server curl python3-pip cloud-init
```

My main.tf file looks very similar to [the simple cloudinit example](https://github.com/Telmate/terraform-provider-proxmox/blob/master/examples/cloudinit_example.tf)
```
variable "SSH_KEY" {
  default = "ssh-ed25519 AAAAC3SNIP"
}
variable "ip_address" {
  default = "SNIP.50"
}
variable "cidr" {
  default = "24"
}
variable "gateway" {
  default = "SNIP.1"
}
variable "nameservers" {
  default = "SNIP SNIP"
}

terraform {
  required_providers {
    proxmox = {
      source = "Telmate/proxmox"
    }
  }
}

provider "proxmox" {
}

resource "proxmox_vm_qemu" "proxmox_vm" {
  count             = 1
  name              = "ns1.DOMAIN"
  target_node       = "SNIP"
  clone             = "ubuntu-20.04"
  full_clone        = false  # Linked clone
  os_type           = "cloud-init"
  cores             = 1
  sockets           = "1"
  cpu               = "host"
  memory            = 512
  scsihw            = "virtio-scsi-pci"
  bootdisk          = "scsi0"
  disk {
    size            = "20G"
    type            = "scsi"
    storage         = "zfsdisk"
    iothread        = true
  }
  network {
    model           = "virtio"
    bridge          = "vmbr0"
  }
  # Cloud Init Settings
  # Reference: https://pve.proxmox.com/wiki/Cloud-Init_Support
  ipconfig0 = "ip=${var.ip_address}/${var.cidr},gw=${var.gateway}"
  nameserver = var.nameservers
  sshkeys = <<EOF
var.SSH_KEY
SNIP
  EOF
  # Connection info for all of the provisioners to use
  connection {
    type = "ssh"
    host = var.ip_address
    #host = "SNIP.207"
    user = "root"
    private_key = file(pathexpand("~/.ssh/id_ed25519"))
  }

  # Provisioners
  provisioner "file" {
    source = "named.conf.options"
    destination = "/etc/bind"
  }

  # More provisioners here
```

When I do a `terraform apply` it clones the VM and starts up just fine, but my provisioner can never SSH into it because the IP address of the VM is never updated to use the value from ipconfig0.  If I changes the connection to have `host = "TEMPLATE_IP_ADDRESS"` it will run the provisioners just fine.

I looked through the code to understand how ipconfig0 is used and it looks like it's being [passed on](https://github.com/Telmate/terraform-provider-proxmox/blob/61a2083f63facaeb38c74fbb75d47c484637ece6/proxmox/resource_vm_qemu.go#L727) to pxapi.ConfigQemu, so maybe this ticket belongs in [the proxmox-api-go repo](github.com/Telmate/proxmox-api-go/proxmox), but it seems unlikely that ipconfig0 would have gotten plumbed through here if it didn't work.  Ergo, I must conclude that there's something wrong with my setup, rather than with the code in either of these repos.  I just can't figure out what that might be.

Now, I can almost certainly hack around this by having the first provisioner override the host in the connection section to be the template IP address, and have that provisioner manually change the ip address so the remaining providers will work, but I'd much rather use the normal tools (cloud-init in this case) if possible.

Any suggestions on how I can troubleshoot this?

Note: Once I get the IP address being set properly, I do plan on having a .yml file as is shown in the [cloud-init guide](https://github.com/Telmate/terraform-provider-proxmox/blob/master/docs/guides/cloud_init.md) as shown below.  If this is going to conflict with using ipconfig0, please let me know.
```
  provisioner "file" {
    source      = "cloud-init.yml"
    destination = "/var/lib/vz/snippets/user_data_vm-${count.index}.yml"
  }
```

**Comment by AcidCrash1337** [comment]
I now have an even greater suspicion this is not a problem with this terraform plugin, as I attempted to clone a template and create a VM using the GUI.  I'm getting very mixed instructions on how to do this.

One one hand, [the proxmox cloud init support page](https://pve.proxmox.com/wiki/Cloud-Init_Support) says "Proxmox VE generates an ISO image to pass the Cloud-Init data to the VM."  Great, I didn't want to mess around with creating ISO9660 images and this also matches what the terraform provider seems to be doing (i.e. just passing values to Proxmox, not generating an ISO9660 image and passing that to Proxmox).

On the other hand, the [Cloud-Init FAQ](https://pve.proxmox.com/wiki/Cloud-Init_FAQ) says the CD-ROM must be added and the "CloudInit Drive" selected.  I don't have a CloudInit drive.  I just have Ubuntu installation ISOs that I uploaded.  I could generate an ISO with the proper name and configuration files in [NoCloud v1 format](https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html) if I have to, but I haven't seen anyone using Proxmox suggest that this is necessary.

Setting ipconfig0 using the instructions for [deploying cloud init image](https://pve.proxmox.com/wiki/Cloud-Init_Support#_deploying_cloud_init_templates) does cause it to appear in the web UI under "Cloud-Init", however it still says No CloudInit Drive found.  Booting this VM gives me the template IP, not the one specified in ipconfig0.  This does match my experience with the terraform provider, that the value is there but it doesn't use it.  When I search online for where I might get this "Cloudinit Drive", I'm coming up empty handed.

Does everyone else have some CloudInit ISO in their local datastore by default?  Is there same way I can obtain said ISO?  I'm using Proxmox VE 6.2-4, and it is fully patched.

Finally, there is cicustom, which appears to be a way to specify a volume that contains the cloud-init yaml configuration file.  If this worked, it seems like it would bypass all of this CD-ROM nonsense.  I followed the instructions in the [Cloud-init guide](https://github.com/Telmate/terraform-provider-proxmox/blob/de7ba2b30ce73cd4aa5f8242f8be452e2c1f8e45/docs/guides/cloud_init.md) and found that my .yml file was not getting uploaded to the proxmox node, nor was it giving me any error about failing to be uploaded.  I can try manually putting a file on the node and referencing it with cicustom, to help diagnose the issue, but I don't understand what a long-term fix would look like here.

If I am able to figure this out, I vow to submit a pull request that includes expanded documentation on how cloud-init works, points to the right reference materials, and explains things in a more digestible way.  If nothing else, it'll serve me well when I forget whatever the solution was a couple months from now and need a refresher. :slightly_smiling_face:

**Comment by AcidCrash1337** [comment]
:astonished: I found the problem!  While the Proxmox docs say the configs are passed via a CD-ROM drive, they do not say that it's a magical CD-ROM that is NOT the "CD/DVD Drive" that is listed on the Hardware section of the web UI.  Thanks to [LearnLinuxTV](https://youtu.be/8qwnXd1yRK4?t=582) for spelling this out for me.  Going back and re-reading the Proxmox documentation, I can see that it does actually say this and I just misread it.  Since I'm unlikely to be the only person to misread this, I'll see if I can update those docs in order to be clearer.  It is a wiki, but it doesn't have any edit functionality, but maybe that will change if I register for an account.

Now, the next question is: if `os_type` is set to "cloud-init" (which it is in my configuration), why didn't the VM have this magic CD-ROM to pass along cloud-init configuration files?

I re-read the both of the configuration files from the example and the guide and neither mentioned anything about adding this CouldInit drive.  However, they are both cloning from VMs that I do not have any information about.  I manually added this device to my base image and tried re-deploying with terraform, and it worked like a charm!

Once I knew what to look for, I checked to see if anyone is talking about this on the Packer issue tracker and sure enough, someone put in a feature request to [allow adding the cloud-init drive](https://github.com/hashicorp/packer/issues/8691) and it was [implemented](https://github.com/hashicorp/packer/pull/9045) in Packer 1.5.6.  Just set `"cloud_init": true` inside the builder section of your packer template.  I tested this out (packer 1.6.5 on Ubuntu 18.04) and verified that it worked as expected.

To the people of the future who read this, I hope you enjoyed the ride and found this information helpful. :+1:
